pixelator.html

```diff
@@ -1,126 +1,172 @@
-
-        for (let row = 1; row < imageHeight; row = row + pixelMultiplier) {
-          for (let col = 1; col < imageWidth; col = col + pixelMultiplier) {
-            loopkiller--;
-            if (loopkiller < 0) { return console.log("killed"); }
-            
-            // get 1st pixel color
-            const pixelReference = ((row*imageWidth)+(col))*4;
-            if (pixelReference === null || typeof pixelReference === "undefined") {return console.log("Invalid pixel reference"); }
-
-            const color = `rgb(
+<head>
+  <meta charset="UTF-8">
+  <meta http-equiv="X-UA-Compatible" content="IE=edge">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <meta name="author" content="Joel Peltonen">
+  <title>Document</title>
+  <style>
+    * {
+      margin: 0;
+      padding: 0;
+    }
+
+    body {
+      display: flex;
+      flex-direction: column;
+      justify-content: center;
+      align-items: center;
+      width: 100vw;
+      height: 100vh;
+    }
+
+    #controls {
+      display: flex;
+      flex-direction: column;
+      justify-self: center;
+      text-align: right;
+      background-color: whitesmoke;
+    }
+
+    .input-group {
+      display: flex;
+      align-items: center;
+      margin: .5em;
+    }
+
+    label {
+      width: 10em;
+      margin-right: .5em;
+    }
+
+    #graphics {
+      display: flex;
+      justify-content: center;
+      align-items: center;
+    }
+  </style>
+</head>
+
+<body onload="initPage()">
+<div id="controls">
+  <div class="input-group">
+    <label for="imageFile">Image File:</label>
+    <input type="file" name="" id="imageFile">
+  </div>
+  <div class="input-group">
+    <label for="multiplierInput">Pixelation Factor:</label>
+    <input type="number" id="multiplierInput" min="0" max="666" value="2" />
+  </div>
+</div>
+<div id="graphics">
+  <img id="sourceImage" src="" alt="source image"/>
+  <canvas id="targetCanvas"></canvas>
+</div>
+Joel Peltonen 2021
+
+<script type="text/javascript">
+
+  const initialFilename = 'Pink-Gonzo.jpg';
+  const maxFileSize = 900000;
+  const maxPixelationFactor = 666;
+  const minPixelationFactor = 2;
+  const initialPixelationFactor = 5;
+  let pixelationFactor = initialPixelationFactor;
+
+  // Are you KIDDING me with this type casting?
+  const canvas = /** @type {HTMLCanvasElement} */ document.querySelector('#targetCanvas');
+  const ctx = /** @type {CanvasRenderingContext2d} */ canvas.getContext('2d');
+  const fileDialog = document.querySelector("#imageFile");
+  const sourceImage = document.querySelector('#sourceImage');
+  const multiplierInput = document.querySelector("#multiplierInput");
+
+  // init page (triggered by body onload event)
+  function initPage() {
+    sourceImage.setAttribute('src', initialFilename);
+    multiplierInput.setAttribute('value', pixelationFactor);
+  }
+
+  // event handlers
+  fileDialog.addEventListener('change', () => {
+    readFile();
+    draw();
```

Comment:
Hmm, actually because there is `sourceImage.addEventListener('load', () => { ..});` here, I don't think that this `draw()` is necessary and it might even be called before an image is loaded. Not 100% sure though, would need to test

Reply:
I am not sure by 100%, too, why I exactly did that. Your code had/has this click-handler for the update button that contains: